### PR TITLE
Support accumulating DDP grads using a context manager

### PR DESCRIPTION
**Pull request by mrshenli**
The first attempt and more discussions are available in #19577 

#### Goal

Allow toggling DDP gradient synchronization across iterations. With this feature, users may accumulate grads in module variables, and only kick off expensive grad synchronize every a few iterations. 

#### Concerns

Our first attempt in #19577 tries to do it using a variable or a function. But @apaszke made a good point that it will not be error prone, and favors a context manager instead.

#### Proposed Solution

Instead of providing a `accumulate_grads` variable/function/context, we provide a `DistributedDataParallel.no_sync()` context manager. And it does exactly what the name suggests, i.e., disable DDP grad synchronization within the context. Note that `accumulate_grads` means `no_sync` + no optimizer step, where the latter is not controlled by DDP. 

It is true that users need to call another `model(input).backward()` after exiting the context, and this is indeed more verbose. But I think it is OK as one major concern in the previous discussion is to prevent users from running into errors without knowing it. This API should reaffirm the expected behavior, and does not mess up with other use cases if accumulating grads is not required.. 

The application would then look like:

```python
with ddp.no_sync():
  for input in inputs:
    ddp(input).backward()

ddp(one_more_input).backward() 
optimizer.step()
```

@chenyangyu1988 @myleott

### CODE REVIEW

**Comment by mrshenli** [comment]
The failed test passed in [rerun](https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-cuda9.0-cudnn7-windows-trigger-build/53475/).

**Review by pietern — COMMENTED** [review]
LGTM in general, but with 2 comments:

1. Would it be possible to consolidate some of the duplication in the test code into some kind of helper class, or a set of helper functions? There is a lot of duplication between these tests and I think there are at least half a dozen more similar to these.
2. I think there is interaction between `no_sync` and the `_sync_params` function that could cause unintended results. For example, the gradients of model replicas are detached and zeroed in every iteration, whereas they should also accumulate. Then there is the question of the batch normalization buffer synchronization in every call to forward... not sure what to do about that one.

**Comment by mrshenli** [comment]
> 1. Would it be possible to consolidate some of the duplication in the test code into some kind of helper class, or a set of helper functions? There is a lot of duplication between these tests and I think there are at least half a dozen more similar to these.

Yes, let me try

> 2. I think there is interaction between no_sync and the _sync_params function that could cause unintended results. For example, the gradients of model replicas are detached and zeroed in every iteration, whereas they should also accumulate. Then there is the question of the batch normalization buffer synchronization in every call to forward... not sure what to do about that one.


Sorry, I forgot about this. How about the following two options:

1. Only allow creating `no_sync` context if the DDP does not contain module replicas, i.e., `_sync_params` becomes no op.

2. The `_sync_params` should be called if the previous iteration invoked `prepare_for_backward`. So, maybe we can add an additional var recording `prepare_for_backward` was invoked last time?

**Review by pietern — APPROVED** [review]
Looks good to me!

I think the new tests didn't fail because it doesn't require any buffer synchronization. If it did, it would have yielded different results without the `require_forward_param_sync` option.

**Comment by mrshenli** [comment]
> I think the new tests didn't fail because it doesn't require any buffer synchronization. If it did, it would have yielded different results without the require_forward_param_sync option.

@pietern do we want to raise an exception if `no_sync` is called from model with buffers?

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -377,8 +380,31 @@ def _check_default_group(self):
...
+
+            >>> ddp = torch.nn.DistributedDataParallel(model, pg)
+            >>> with ddp.no_sync():
+            ...   for input in inputs:
+            ...     ddp(input).backward()  # no synchronization, accumulate grads
+            ... ddp(another_input).backward()  # synchronize grads
+        """
+        old_require_backward_grad_sync = self.require_backward_grad_sync
+        self.require_backward_grad_sync = False
+        yield
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: This should be in a `try ... finally` block, because in case of an exception you will fail to restore the flag!

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -272,6 +273,8 @@ def __init__(self, module, device_ids=None,
         self.module = module
         self.broadcast_buffers = broadcast_buffers
         self.find_unused_parameters = find_unused_parameters
+        self.require_backward_grad_sync = True
+        self.require_forward_param_sync = True
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: I'm not sure if `DistributedDataParallel` is picklable, but if it is, then you should add a `__setstate__` that adds those two attributes, because otherwise people who load older checkpoints will get missing attribute errors.

**Comment by apaszke** [comment]
I've fixed the problems in a PR that's referenced above.
